packages/tasit-action/src/TasitAction.js

```diff
@@ -19,117 +22,370 @@ class Utils {
+        );
+        return false;
+      }
+      return true;
+    });
+  };
+
+  off = (eventName, listener) => {
+    this.removeListener(eventName, listener);
+  };
+
+  listenerCount = eventName => {
+    return this.#events.filter(event => {
+      return event.eventName === eventName;
+    }).length;
+  };
+
+  removeAllListeners = () => {
+    this.#events.forEach(event => {
+      this.removeListener(event.eventName, event.listener);
+    });
+  };
+
+  unsubscribe = () => {
+    this.removeAllListeners();
+  };
+
+  // TODO: Make protected
+  emitErrorEvent = error => {
+    this.#events.forEach(event => {
+      if (event.eventName === "error") {
+        event.listener(error);
+      }
+    });
+  };
+
+  // TODO: Make protected
+  addErrorListener = listener => {
+    const eventName = "error";
+    this.#events.push({
+      eventName,
+      wrappedEventName: eventName,
+      listener,
+      wrappedListener: listener,
+    });
+  };
+
+  // TODO: Make protected
+  getEventInfo = (wrappedEventName, wrappedListener) => {
+    return this.#events.find(
+      event =>
+        event.wrappedEventName === wrappedEventName &&
+        event.wrappedListener === wrappedListener
+    ).info;
+  };
+
+  // TODO: Make protected
+  setEventInfo = (wrappedEventName, wrappedListener, info) => {
+    this.#events.forEach(event => {
+      if (
+        event.wrappedEventName === wrappedEventName &&
+        event.wrappedListener === wrappedListener
+      )
+        event.info = info;
+    });
+  };
+
+  // TODO: Make protected
+  addListener = (eventName, wrappedEventName, listener, wrappedListener) => {
+    this.#emitter.on(wrappedEventName, wrappedListener);
+
+    this.#events.push({
+      eventName,
+      wrappedEventName,
+      listener,
+      wrappedListener,
+    });
+
+    setTimeout(() => {
+      this.emitErrorEvent(new Error(`Listener removed after reached timeout`));
+      this.removeListener(eventName, listener);
+    }, EVENTS_CONFIG.timeout);
+  };
+}
+
+class TransactionSubscription extends Subscription {
   #txPromise;
   #provider;
+  #tx;
 
   constructor(txPromise, provider) {
+    super(provider);
     this.#txPromise = txPromise;
     this.#provider = provider;
   }
 
-  on = async (trigger, callback) => {
-    if (trigger !== "confirmation") {
-      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+  on = async (eventName, listener) => {
+    const triggers = ["confirmation", "error"];
+
+    if (!triggers.includes(eventName)) {
+      throw new Error(`Invalid listener trigger, use: [${triggers}]`);
     }
 
-    if (!callback || typeof callback !== "function") {
-      throw new Error(`Cannot subscribe without a function`);
+    if (!listener || typeof listener !== "function") {
+      throw new Error(`Cannot listening without a function`);
     }
 
-    const blockHeight = 1;
-    const tx = await this.#txPromise;
+    if (eventName === "error") {
+      this.addErrorListener(listener);
```

Comment:
Sorry, I didn't get it.

Reply:
Oh, what I meant is that the `addErrorListener` function doesn't do a whole lot of business logic beyond pushing the event and listener (it does wrap them though, which is something).

I should have made the comment where the `addErrorListener` function itself is defined to be clearer.

If we ended up having a lot of events named error with different payloads, we might want to parse the payload in this function or something.

I don't think there's an immediate action item here. Let's just keep the topic in mind as this file keeps evolving.

Reply:
Since there's no immediate action item here, this thread is a low priority.

However, @marcelomorgado reading it quickly and then forgetting about it until it comes up again is a high priority while it's fresh in our minds.

Reply:
I'm resolving this since you gave that last comment a thumbs-up